// @workadventure/tiled-map-type-guard/dist/ITiledMapTileLayer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ITiledMapTileLayer = void 0;
const zod_1 = require("zod");
const ITiledMapProperty_1 = require("./ITiledMapProperty");
const ITiledMapChunk_1 = require("./ITiledMapChunk");
exports.ITiledMapTileLayer = zod_1.z.object({
    data: zod_1.z.union([zod_1.z.string(), zod_1.z.number().array()]),
    height: zod_1.z.number(),
    id: zod_1.z.number(),
    name: zod_1.z.string(),
    opacity: zod_1.z.number(),
    type: zod_1.z.literal('tilelayer'),
    visible: zod_1.z.boolean(),
    width: zod_1.z.number(),
    chunks: ITiledMapChunk_1.ITiledMapChunk.array().optional(),
    compression: zod_1.z.string().optional(),
    encoding: zod_1.z.enum(['csv', 'base64']).optional(),
    offsetx: zod_1.z.number().optional(),
    offsety: zod_1.z.number().optional(),
    parallaxx: zod_1.z.number().optional(),
    parallaxy: zod_1.z.number().optional(),
    properties: ITiledMapProperty_1.ITiledMapProperty.array().optional(),
    startx: zod_1.z.number().optional(),
    starty: zod_1.z.number().optional(),
    tintcolor: zod_1.z.string().optional(),
    class: zod_1.z.string().optional(),
    x: zod_1.z.number().optional(),
    y: zod_1.z.number().optional(),
});
